omroneipApp/src/drvOmroneip.cpp

```diff
@@ -0,0 +1,1882 @@
+      return structName;
+    }
+    else {
+      return "structName is neither a structure or a standard dtype";
+    }
+  }
+  return "no dtypes left";
+}
+
+void drvOmronEIP::readPoller()
+{
+  static const char *functionName = "readPoller";
+  std::string threadName = epicsThreadGetNameSelf();
+  std::string tag;
+  int offset;
+  int status;
+  int still_pending = 1;
+  int timeTaken = 0;
+  //The poller may not be fully initialised until the startPollers_ flag is set to 1, do not attempt to get the pPoller yet
+  while (!this->startPollers_ && !omronExiting_) {epicsThreadSleep(0.1);}
+  omronEIPPoller* pPoller = pollerList_.at(threadName);
+  double interval = pPoller->updateRate_;
+  asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Starting poller: %s with interval: %f\n", driverName, functionName, threadName.c_str(), interval);
+  while (!omronExiting_)
+  {
+    double waitTime = interval - ((double)timeTaken / 1000);
+    if (waitTime >=0) {
+      epicsThreadSleep(waitTime);
+    }
+    else {
+      asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Warning! Reads taking longer than requested! %f > %f\n", driverName, functionName, ((double)timeTaken / 1000), interval);
+    }
+
+    auto startTime = std::chrono::system_clock::now();
+    for (auto x : tagMap_)
+    {
+      if (x.second->pollerName == threadName && x.second->readFlag == true)
+      {
+        asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Reading tag: %d with polling interval: %f seconds\n", driverName, functionName, x.second->tagIndex, interval);
+        plc_tag_read(x.second->tagIndex, 0); // read from plc as fast as possible, we will check status and timeouts later
+      }
+    }
+
+    int waits = 0;
+    bool readFailed = false;
+    for (auto x : tagMap_)
+    {
+      if (x.second->pollerName == threadName)
+      {
+        offset = x.second->tagOffset;
+        still_pending = 1;
+        
+        while (still_pending)
+        {
+          // It should be rare that data for the first tag has not arrived before the last tag is read
+          // Therefor this loop should rarely be entered and would only be called by the first few tags as we 
+          // are asynchronously waiting for all tags in this poller to be read. 
+          still_pending = 0;
+          status = plc_tag_status(x.second->tagIndex);
+
+          if (status == PLCTAG_STATUS_PENDING)
+          {
+            // Wait for the timeout specified in the record
+            if (waits*0.01>=x.second->timeout)
+            {
+              setParamStatus(x.first, asynTimeout);
+              readFailed = true;
+              asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Timeout finishing read tag %d: %s. Try decreasing the polling rate\n", driverName, functionName, x.second->tagIndex, plc_tag_decode_error(status));
+              break;
+            }
+            else
+            {
+              still_pending = 1;
+              waits++;
+              epicsThreadSleep(0.01);
```

Comment:
Why is this sleep 0.01 needed here?

Reply:
I have essentially implemented by own timer here. I dont want to progress until a timeout has expired and am using a combination of counting up waits and epicsThreadSleep to count up to the timeout. I think I did it this way rather than allowing the thread to constantly loop and checking the actual time, as I thought it was useful to sleep the thread. This was maybe back when the thread was locking? Anyway, I will switch it to look at a clock which is more intuitive

Reply:
I have changed it to use a std::chrono timer and I think it should be fine. It is a bit complex as there is both the timing interval for the poller and the timeout for the record. We send all the read requests at the start. This loop we are looking at is waiting for timeouts if the status we recieve from libplctag is PLCTAG_STATUS_PENDING. It will wait until this records timeout is expired and then skip to the next item in the poller. Hopefully this item will now have been fetched, if not then if its timeout was the same as the previous record, it will now have expired etc. If the user sets timeouts so that they extend beyond the read interval requested, then this is reported, however we still wait for the timeouts and do not abort this poll.

Reply:
I have just pushed the changes now